### ISSUE TITLE
Ideas on improving Crate Validator

### ISSUE THREAD

**Issue by Cardosaum** [risc0/risc0#835]
# Overview

We propose incremental changes to the already existing tools Crate Validator  and `cargo risczero test`, as well as a revamp on the report UI; This changes seek to remove overlaping functionalities among the tools, and to provide a more user-friendly experience to developers seeking to use the zkVM.

## Motivation

Our current Crate Validator tool allow users to check if crates listed in crates.io are compatible with our zkVM, giving them the flexibility to query individual or a list of crates. However, the current tool is not user-friendly and provides no support for selectively enabling/disabling features for the requested crate, nor to run the test against a pre-specified crate version (always checking the latest one). 

Also, with the incoming [`cargo risczero test`][1] command, we will have two tools with overlaping functionalities around the building process of specified crates.

Another issue that we would seek to address is an improvement on the current report UI, allowing users to easily identify issues found during crate compilation steps, and to provide more detail around current and previous status for each crate.

In the same topic of report UI, we'd like to provide a more stable version of the report, avoiding situations where the report is not available due to a failure on the build process of a crate. Such an issue occured recently with the `parity-scale-codec-derive` crate, which was failing to build due to a missing dependency (`cargo risczero install`) during CI setup, and was causing the report to show all crates as failing to build.

## Proposed changes

Taking into account the issues mentioned above, we propose the following changes:

### Crate Validator

In order to remove overlaping functionalities between Crate Validator and `cargo risczero test`, we propose to use Crate Validator as an orchestration tool, which will have the following responsibilities:

- Perform selection of crates to check based on user input
- Existing build logic present in Crate Validator will be removed, and replaced by a call to `cargo risczero test` with the appropriate flags.
- Data from build and test steps will be exported enabling posterior usage by the report service that will consume it in order to present up-do-date information in a user-friendly way.

One change that we propose to the current Crate Validator is to allow users to specify a `Cargo.toml` file instead of a custom JSON. By doing this, users can specify their dependencies in the way they're already used to do in normal rust project, giving them the flexibility to specify crate version, patches, features, etc. 

### `cargo risczero test`

The `cargo risczero test` command will be responsible for building and testing the specified crate, will be pretty much unchanged from the current implementation, only needing to be added to a wrapping call inside the Crate Validator.

### Report service

One possibility is that we implement a simple REST API that will be responsible for receiving compilation and test results from the Crate Validator and presenting it in a user-friendly way, as demonstrated in the image below.

Maybe we could also provide a way for users to request a report given a specific `Cargo.toml` definition? This would allow users to offload the compilation and testing process to the report service, and to receive a report with the results.

#### Proposed initial service UI: 

![Service UI](https://user-images.githubusercontent.com/45436839/264992526-a4f03f3d-ec48-4b73-8fe4-3e4625e1b770.svg)

[1]: https://github.com/risc0/risc0/pull/646

## Observations

It's important to note that some of the features here already exists both in the Crates Validator and report; We would do incremental changes to the existing tools, and would not need to start from scratch.

Notable features missing are the usage of `Cargo.toml` file instead of a custom JSON in the Crate Validator, and the `cargo risczero test` command, which is still in development and might change before being merged.

With regards to the report service, all the REST API and respective UI would need to be implemented from scratch, but we could reuse the existing report UI as a starting point.

**Comment by Cardosaum** [comment]
Curious if some of you would have inputs on this matter. 

cc/ @mothran  @carterbrett @nategraf @jjtny1

**Comment by nategraf** [comment]
I think thinking about the crate validator as an orchestration tool, and one specifically designed to generate the compatibility reports is the right direction. Having it use `cargo risczero build` and `cargo risczero test` internally seems correct to me as well. 👍 

I like the idea of generating a standardized output from the core tool e.g. in CSV or JSON format and parsing that in the UI that displays the page. Note that we can make this as simple as uploading the reports to an S3 bucket and serving a static HTML / TS webpage that parses the CSV or JSON report and produces the visuals. With this there is essentially no continually running service overhead, just a batch pipeline and some static resources.

I don't think we should be making an effort to make user-requested reports. Instead, when a user wants to know if a a crate will build for the RISC Zero target, they should simply build it using `cargo risczero build`. This is much simpler and more direct method for developers, and does not require us to run a service that takes user requests.

Instead I think the tool should focus on periodic reporting, running builds and tests for a list of crates and versions nightly. Users can request new crates, for example crates they wrote, be added to the report and we could accommodate that via pull request or via a Google form. Users can also report crates that _don't_ build with an issue to `risc0` and we should provide an issue template for that.

I also don't think using `Cargo.toml` to configure the crate validator tool is the right approach. The crate validator tool is an orchestration tool, and wraps the cargo build system. It's configuration goes well beyond what cargo is designed to support. For example, we may want to test a matrix of a crate's versions and features and for that we would likely want our own configuration schema. Build profiles are another example of this where we want to encode additional steps required to get a crate to build. Additionally, when we start running unit tests with `cargo risczero test` we will likely want to add test filters, which is another configuration not present in `Cargo.toml`

Instead I think we should have a YAML or TOML configuration for the crate validator. Over time we will add more crates and configurations to it to test, and it provides a location for library developers to make their pull requests.

One feature that would be nice to support in the future is to have our build system check the dependency graph for crates with known issues, and especially known workarounds for these issues. This should definitely be included in the crates validator report, and if it is posted to a well known location the build system could fetch it and use the data included. This would greatly increase discoverability of this tool and allow developers to use the known workarounds quicker instead of getting stuck.

Thanks for working on this tool! It's going to be super valuable for developers as people implement more and more complicated applications on top of the zkVM.

**Comment by Cardosaum** [comment]
Thanks for the feedback, Victor!
I agree with you on all points, but I have some questions:

1. When you mention the addition of filters for tests, you mean running only specific tests, or tests that match a specific pattern, right?

2. On the dependency graph check, how are you envisioning it to work? I'd imagine that we'd scan the dependency graph and see if there is any crate that we know won't compile or the ones that need a workaround, but how exactly should we proceed? For the ones that do have a workaround, should we, during compilation apply the workaround automatically (if so, how)? Or should we just warn the user that there is a workaround for that crate and let them decide if they want to apply it or not (also giving instructions on how they should proceed in applying it)?

**Comment by nategraf** [comment]
> When you mention the addition of filters for tests, you mean running only specific tests, or tests that match a specific pattern, right?

That's right. By default we should try running all the tests, but I imagine there will be a lot of cases where many of tests either don't work (e.g. they require filesystem access) or are too time consuming (e.g. fuzzing tests with high iteration counts).
